File: app/lib/services/NoteService.ts
import { NoteRepository } from '../repositories/NoteRepository';
import { Note, NoteWithStats, NoteCreateRequest, PaginationParams, PaginatedResponse } from '../../types';

export class NoteService {
  private noteRepository: NoteRepository;

  constructor() {
    this.noteRepository = new NoteRepository();
  }

  async createNote(noteData: NoteCreateRequest, authorId: number): Promise<number> {
    if (!noteData.title || !noteData.content) {
      throw new Error('标题和内容不能为空');
    }

    return this.noteRepository.create(noteData, authorId);
  }

  async updateNote(noteId: number, noteData: Partial<NoteCreateRequest>, authorId: number): Promise<boolean> {
    // 验证笔记属于当前用户
    const note = await this.noteRepository.findById(noteId);
    if (!note || note.author_id !== authorId) {
      throw new Error('笔记不存在或无权修改');
    }

    return this.noteRepository.update(noteId, noteData);
  }

  async getNoteById(noteId: number): Promise<NoteWithStats | null> {
    return this.noteRepository.findByIdWithStats(noteId);
  }

  async getUserNotes(userId: number, pagination: PaginationParams): Promise<PaginatedResponse<NoteWithStats>> {
    return this.noteRepository.findByAuthor(userId, pagination);
  }

  async getPublicNotes(pagination: PaginationParams): Promise<PaginatedResponse<NoteWithStats>> {
    return this.noteRepository.findPublicNotes(pagination);
  }

  async searchPublicNotes(query: string, pagination: PaginationParams): Promise<PaginatedResponse<NoteWithStats>> {
    if (!query || query.trim().length < 2) {
      throw new Error('搜索关键词至少需要2个字符');
    }

    return this.noteRepository.searchPublicNotes(query.trim(), pagination);
  }

  async deleteNote(noteId: number, authorId: number): Promise<boolean> {
    // 验证笔记属于当前用户
    const note = await this.noteRepository.findById(noteId);
    if (!note || note.author_id !== authorId) {
      throw new Error('笔记不存在或无权删除');
    }

    return this.noteRepository.delete(noteId);
  }

  async getNoteStats(noteId: number): Promise<{
    likeCount: number;
    bookmarkCount: number;
    commentCount: number;
  }> {
    const note = await this.noteRepository.findByIdWithStats(noteId);
    if (!note) {
      throw new Error('笔记不存在');
    }

    return {
      likeCount: note.likeCount || 0,
      bookmarkCount: note.bookmarkCount || 0,
      commentCount: note.commentCount || 0
    };
  }
}

File: partjava-backend/src/main/java/com/partjava/service/NoteService.java
package com.partjava.service;

import com.partjava.entity.Note;
import java.util.List;

public interface NoteService {
    /**
     * 条件筛选列表获取笔记
     */
    List<Note> listNotes(Long userId, String category, String technology, Boolean isPublic);

    /**
     * 获取单个笔记详情（外部调用，无用户上下文）
     */
    Note getNoteById(Long noteId);

    /**
     * 获取单个笔记详情（带用户鉴权，非公开笔记仅作者可查看）
     */
    Note getNoteById(Long noteId, Long userId);

    /**
     * 新建随堂笔记
     */
    Note createNote(Long userId, Note note);

    /**
     * 修改随堂笔记
     */
    Note updateNote(Long userId, Long noteId, Note note);

    /**
     * 删除随堂笔记
     */
    void deleteNote(Long userId, Long noteId);

    /**
     * 一键点赞或取消点赞笔记
     */
    void toggleLikeNote(Long userId, Long noteId);

    /**
     * 一键收藏或取消收藏笔记
     */
    void toggleBookmarkNote(Long userId, Long noteId);

    /**
     * 进阶会员专享：一键复用/克隆他人公开笔记
     */
    Note cloneNote(Long userId, Long noteId);
}
